lib/optional_thunk.ml

```diff
@@ -0,0 +1,11 @@
+type t = unit -> unit
+
+let none = Sys.opaque_identity (fun () -> ())
+let some f =
+  if f == none
+  then failwith "Optional_thunk: this function is not representable as a some value";
+  f
+
+let is_none t = t == none
+let is_some t = not (is_none t)
+let unchecked_value t = t
```

Comment:
I was thinking about this when I was reviewing it and asking myself "should this be raising if `t` is `none`?" – but that felt a little burdensome in all of the places we use it safely.

But now I'm thinking about it some more, and I'm wondering: maybe it would make sense if we split this into two functions, kind of like what we talked about before the PR:

```ocaml
(** Return the exposed thunk, raises if called on [none] *)
val value_exn : t -> unit -> unit

(** Invoke the thunk if [some], do nothing if [none] *)
val invoke_if_some : t -> unit
```

This has two benefits: calling an optional thunk should always be safe, since the default is to just noop by invoking `none`, and we can use this in all of the current places where we use `unchecked_value` without `is_some`. But removing opaqueness is something you have to be a bit more careful about, since presumably you should be able to do `Optional_thunk.some (Optional_thunk.value_exn t)`. And ultimately I rank "descriptive names" below "failing at runtime" and "compiler error" in terms of safety.